In [1]:
import pandas as pd
import numpy as np
from pathlib import Path

In [2]:
PROJECT_ROOT = Path.cwd().parent
RAW_DATA = PROJECT_ROOT / "data" / "raw" / "PS_20174392719_1491204439457_log.csv"

print(RAW_DATA)
print(f"Exist:{RAW_DATA.exists()}")

C:\Users\User\Desktop\fintech-transaction-performance-risk-analytics\data\raw\PS_20174392719_1491204439457_log.csv
Exist:True


In [3]:
df = pd.read_csv(RAW_DATA)

In [4]:
df.shape

(6362620, 11)

In [5]:
df.info()


<class 'pandas.DataFrame'>
RangeIndex: 6362620 entries, 0 to 6362619
Data columns (total 11 columns):
 #   Column          Dtype  
---  ------          -----  
 0   step            int64  
 1   type            str    
 2   amount          float64
 3   nameOrig        str    
 4   oldbalanceOrg   float64
 5   newbalanceOrig  float64
 6   nameDest        str    
 7   oldbalanceDest  float64
 8   newbalanceDest  float64
 9   isFraud         int64  
 10  isFlaggedFraud  int64  
dtypes: float64(5), int64(3), str(3)
memory usage: 534.0 MB


In [6]:
df.head()

,step,type,amount,nameOrig,oldbalanceOrg,newbalanceOrig,nameDest,oldbalanceDest,newbalanceDest,isFraud,isFlaggedFraud
0,1,PAYMENT,9839.64,C1231006815,170136.0,160296.36,M1979787155,0.0,0.0,0,0
1,1,PAYMENT,1864.28,C1666544295,21249.0,19384.72,M2044282225,0.0,0.0,0,0
2,1,TRANSFER,181.00,C1305486145,181.0,0.00,C553264065,0.0,0.0,1,0
3,1,CASH_OUT,181.00,C840083671,181.0,0.00,C38997010,21182.0,0.0,1,0
4,1,PAYMENT,11668.14,C2048537720,41554.0,29885.86,M1230701703,0.0,0.0,0,0


In [7]:
print("Columns:")
print(df.columns.tolist())

print("\nFirst 5 rows:")
display(df.head())

Columns:
['step', 'type', 'amount', 'nameOrig', 'oldbalanceOrg', 'newbalanceOrig', 'nameDest', 'oldbalanceDest', 'newbalanceDest', 'isFraud', 'isFlaggedFraud']

First 5 rows:


,step,type,amount,nameOrig,oldbalanceOrg,newbalanceOrig,nameDest,oldbalanceDest,newbalanceDest,isFraud,isFlaggedFraud
0,1,PAYMENT,9839.64,C1231006815,170136.0,160296.36,M1979787155,0.0,0.0,0,0
1,1,PAYMENT,1864.28,C1666544295,21249.0,19384.72,M2044282225,0.0,0.0,0,0
2,1,TRANSFER,181.00,C1305486145,181.0,0.00,C553264065,0.0,0.0,1,0
3,1,CASH_OUT,181.00,C840083671,181.0,0.00,C38997010,21182.0,0.0,1,0
4,1,PAYMENT,11668.14,C2048537720,41554.0,29885.86,M1230701703,0.0,0.0,0,0


In [8]:
df["type"].value_counts()

type
CASH_OUT    2237500
PAYMENT     2151495
CASH_IN     1399284
TRANSFER     532909
DEBIT         41432
Name: count, dtype: int64

In [9]:
df["type"].value_counts(normalize=True).mul(100).round(2)

type
CASH_OUT    35.17
PAYMENT     33.81
CASH_IN     21.99
TRANSFER     8.38
DEBIT        0.65
Name: proportion, dtype: float64

In [11]:
print(f"Minimum step: {df['step'].min()}")
print(f"Maximum step: {df['step'].min()}")
print(f"Unique steps: {df['step'].nunique()}")

Minimum step: 1
Maximum step: 1
Unique steps: 743


In [12]:
df["step"]. value_counts().sort_index().tail()

step
739    10
740     6
741    22
742    14
743     8
Name: count, dtype: int64

In [13]:
print("Min:", df["step"].min())
print("Max:", df["step"].max())
print("Unique:", df["step"].nunique())
print("Missing step values:", 744 - df["step"].nunique())

Min: 1
Max: 743
Unique: 743
Missing step values: 1


In [14]:
expected_steps = set(range(1, 744))
actual_steps = set(df["step"].unique())

missing_steps = sorted(expected_steps - actual_steps)

print("Missing steps:", missing_steps)
print("Number missing:", len(missing_steps))

Missing steps: []
Number missing: 0


In [15]:
df.isnull().sum()

step              0
type              0
amount            0
nameOrig          0
oldbalanceOrg     0
newbalanceOrig    0
nameDest          0
oldbalanceDest    0
newbalanceDest    0
isFraud           0
isFlaggedFraud    0
dtype: int64

In [16]:
df.duplicated().sum()

np.int64(0)

In [19]:
print(df["isFraud"].value_counts())
print()
print(df["isFraud"].value_counts(normalize=True).mul(100).round(4))

isFraud
0    6354407
1       8213
Name: count, dtype: int64

isFraud
0    99.8709
1     0.1291
Name: proportion, dtype: float64


In [21]:
print(df["isFlaggedFraud"].value_counts())
print()
print(df["isFlaggedFraud"].value_counts(normalize=True).mul(100).round(4))

isFlaggedFraud
0    6362604
1         16
Name: count, dtype: int64

isFlaggedFraud
0    99.9997
1     0.0003
Name: proportion, dtype: float64


In [22]:
type_counts = df["type"].value_counts()

print(type_counts)

type
CASH_OUT    2237500
PAYMENT     2151495
CASH_IN     1399284
TRANSFER     532909
DEBIT         41432
Name: count, dtype: int64


In [23]:
type_share = df["type"].value_counts(normalize=True).mul(100).round(2)

print(type_share)

type
CASH_OUT    35.17
PAYMENT     33.81
CASH_IN     21.99
TRANSFER     8.38
DEBIT        0.65
Name: proportion, dtype: float64


In [24]:
print("Negative amounts:", (df["amount"] < 0).sum())
print("Zero amounts:", (df["amount"] == 0).sum())
print("Negative origin balances:", (df["oldbalanceOrg"] < 0).sum())
print("Negative destination balances:", (df["oldbalanceDest"] < 0).sum())

Negative amounts: 0
Zero amounts: 16
Negative origin balances: 0
Negative destination balances: 0


In [25]:
df[df["amount"] == 0][
    ["step", "type", "amount", "oldbalanceOrg", "newbalanceOrig",
     "oldbalanceDest", "newbalanceDest", "isFraud", "isFlaggedFraud"]
]
df[df["amount"] == 0]["type"].value_counts()

type
CASH_OUT    16
Name: count, dtype: int64

In [28]:
pd.crosstab(
    df["type"],
    df["isFraud"],
    margins=True
)

pd.crosstab(
    df["type"],
    df["isFraud"],
    normalize="index"
).mul(100).round(4)

isFraud,0,1
type,,
CASH_IN,100.0000,0.0000
CASH_OUT,99.8160,0.1840
DEBIT,100.0000,0.0000
PAYMENT,100.0000,0.0000
TRANSFER,99.2312,0.7688


In [29]:
print("Origin balance changes:")
print((df["oldbalanceOrg"] - df["amount"] - df["newbalanceOrig"]).describe())

print("\nDestination balance changes:")
print((df["oldbalanceDest"] + df["amount"] - df["newbalanceDest"]).describe())

Origin balance changes:
count    6.362620e+06
mean    -2.010925e+05
std      6.066505e+05
min     -9.244552e+07
25%     -2.496411e+05
50%     -6.867726e+04
75%     -2.954230e+03
max      1.000000e-02
dtype: float64

Destination balance changes:
count    6.362620e+06
mean     5.556717e+04
std      4.415288e+05
min     -7.588573e+07
25%      0.000000e+00
50%      3.500490e+03
75%      2.935305e+04
max      1.319123e+07
dtype: float64


# Initial Data Profiling Summary

## Dataset Overview

The PaySim dataset contains **6,362,620 transactions across 11 columns**.

The dataset contains transaction-level information covering:
- Transaction timing (`step`)
- Transaction type
- Transaction amount
- Origin and destination accounts
- Origin and destination balances
- Fraud labels
- Fraud detection flags

## Initial Data Quality Findings

| Check | Finding | Status |
|---|---:|---|
| Rows | 6,362,620 | ✅ |
| Columns | 11 | ✅ |
| Missing values | 0 | ✅ |
| Duplicate rows | 0 | ✅ |
| Negative transaction amounts | 0 | ✅ |
| Negative balances | 0 | ✅ |
| Zero-value transactions | 16 | ⚠️ Investigate |
| Observed steps | 1–743 | ✅ |
| Missing steps within observed range | 0 | ✅ |

## Transaction Mix

The dataset contains five transaction types:

- CASH_OUT — 35.17%
- PAYMENT — 33.81%
- CASH_IN — 21.99%
- TRANSFER — 8.38%
- DEBIT — 0.65%

CASH_OUT and PAYMENT together account for approximately **68.98% of all transactions**.

## Fraud Profile

There are **8,213 transactions labeled as fraudulent**, representing approximately **0.1291%** of all transactions.

The simulated fraud-detection flag is present on only **16 transactions**.

Fraud labels are observed in:
- CASH_OUT — 0.1840%
- TRANSFER — 0.7688%

No fraud labels are observed in CASH_IN, DEBIT, or PAYMENT transactions in this dataset.

These are descriptive observations only. Further validation is required before interpreting them as business risk patterns.

## Data Quality Observations Requiring Further Investigation

### 1. Zero-value transactions

There are 16 transactions with an amount of zero. All 16 are CASH_OUT transactions.

These transactions have not been removed or modified. Their behavior will be investigated during the data-quality validation stage.

### 2. Balance reconciliation

Initial balance reconciliation shows that origin and destination balances do not consistently reconcile to the transaction amount.

This is an important validation finding because balance movements are central to understanding transaction behavior and financial integrity.

The discrepancies will **not be corrected during initial profiling**.

They will be investigated during the next stage by:
- transaction type
- fraud status
- account behavior
- balance patterns
- transaction-level examples

## Analytical Principle

> **Profile first. Validate second. Transform third. Analyze fourth. Visualize last.**

Initial profiling identifies patterns and potential data-quality issues. It does not establish causality or justify modifying the source data.

## Conclusion

The dataset is structurally usable for further analysis, with no missing values, duplicate rows, negative transaction amounts, or negative balances identified during initial profiling.

However, several areas require deeper validation before analytical modeling:

1. Balance reconciliation
2. Zero-value transactions
3. Fraud-label and fraud-flag behavior
4. Transaction-type-specific data behavior
5. Account-level transaction patterns

These issues will be investigated before transformation and KPI development.